terraform/lib/dependabot/terraform/update_checker.rb

```diff
@@ -65,34 +67,35 @@ def latest_version_for_registry_dependency
 
         return @latest_version_for_registry_dependency if @latest_version_for_registry_dependency
 
-        versions = all_registry_versions
+        versions = all_module_versions
         versions.reject!(&:prerelease?) unless wants_prerelease?
         versions.reject! { |v| ignore_requirements.any? { |r| r.satisfied_by?(v) } }
 
         @latest_version_for_registry_dependency = versions.max
       end
 
-      def all_registry_versions
-        hostname = dependency_source_details.fetch(:registry_hostname)
-        identifier = dependency_source_details.fetch(:module_identifier)
+      def latest_version_for_provider_dependency
+        return unless provider_dependency?
 
-        # TODO: Implement service discovery for custom registries
-        return [] unless hostname == "registry.terraform.io"
+        return @latest_version_for_provider_dependency if @latest_version_for_provider_dependency
 
-        url = "https://registry.terraform.io/v1/modules/"\
-              "#{identifier}/versions"
+        versions = all_provider_versions
+        versions.reject!(&:prerelease?) unless wants_prerelease?
+        versions.reject! { |v| ignore_requirements.any? { |r| r.satisfied_by?(v) } }
 
-        response = Excon.get(
-          url,
-          idempotent: true,
-          **SharedHelpers.excon_defaults
-        )
+        @latest_version_for_provider_dependency = versions.max
+      end
 
-        raise "Response from registry was #{response.status}" unless response.status == 200
+      def all_module_versions
+        hostname = dependency_source_details.fetch(:registry_hostname)
+        identifier = dependency_source_details.fetch(:module_identifier)
+        RegistryClient.new(hostname: hostname).all_module_versions(identifier: identifier)
+      end
 
-        JSON.parse(response.body).
-          fetch("modules").first.fetch("versions").
-          map { |release| version_class.new(release.fetch("version")) }
+      def all_provider_versions
+        hostname = dependency_source_details.fetch(:registry_hostname)
+        identifier = dependency_source_details.fetch(:module_identifier)
+        RegistryClient.new(hostname: hostname).all_provider_versions(identifier: identifier)
```

Comment:
We could "reuse" this client, but I don't think we actually call it more than twice, so it would just DRY up the code a tiny bit